### ISSUE TITLE
Is there a good way to make sure that multi-row insert with Record operates on the same set of columns

### ISSUE THREAD

**Issue by ethanttbui** [jOOQ/jOOQ#11135]
Suppose I have the following Kotlin code snippet:
```
val record1 = dsl.newRecord(TABLE, mapOf(
    "field1" to "value1",
    "field2" to "value2"
))
val record2 = dsl.newRecord(TABLE, mapOf(
    "field1" to "value1"
))
dsl.insertInto(TABLE)
    .set(record1)
    .newRecord()
    .set(record2)
    .execute();
```
Since `record1` and `record2` do not have the same list of changed fields, for `record2`, jOOQ will infer a `null` value on `field2` upon the insert. In most cases, this is undesirable, as it can lead to unexpected errors, or worse - incorrect data in DB.

Does JOOQ provide an option to change this behavior?

If not, I would like to implement it by myself. But since there's not a `getChangedFields` method, I will need to loop through all fields for each record, call the `Record.changed(field)` method to check, and finally, compare the list of changed fields between all records to make sure that they are the same. The implementation would be simpler with a map, but I'd prefer using Record for easy conversion to/from POJOs.

Is there a better way to implement this?

### Versions:

- jOOQ: 3.14.1
- Java: jdk 1.8
- Database (include vendor): MySQL 5.7.10 InnoDB

**Comment by lukaseder** [comment]
> In most cases, this is undesirable, as it can lead to unexpected errors, or worse - incorrect data in DB.

In most cases, there is no `DEFAULT` expression on the column, in case of which `DEFAULT` and `NULL` semantics tend to be the same, so I'd bargain that the current behaviour is desirable in most cases. Your case is different, of course.

> Does JOOQ provide an option to change this behavior?

Well, what would the changed behaviour look like? In a few RDBMS, we could generate a `DEFAULT` expression instead of a `NULL` value, but that can't be implemented consistently across all dialects.

The guarantee / expectation of this producing a single SQL statement is stronger than the guarantee of only including changed values.

> But since there's not a `getChangedFields` method

I've just recently thought about such a method. Will think about it again: https://github.com/jOOQ/jOOQ/issues/11147. I don't have a clear picture of what the most useful API for this will be. After all, your own implementation could be quite simple:

```java
boolean[] changed = new boolean[record.size()];
for (int i = 0; i < record.size(); i++)
  changed[i] = record.changed(i);
```

Or, use a `BitSet`, or whatever data structure is most appropriate to you.

> Is there a better way to implement this?

You can insert the records 1 by 1. Or you can use `DSLContext.batchInsert()`, which creates several batches for consecutive records that have the same changed flag bitset. Batching single row inserts, or bulk inserting is not the same thing, but it might be good enough for you.

**Comment by ethanttbui** [comment]
Thank you for your answer! Your suggested implementation looks neat, I will give it a try.

> Well, what would the changed behaviour look like?
> The guarantee / expectation of this producing a single SQL statement is stronger than the guarantee of only including changed values.

While in terms of API interface, the current design is good, I would argue that it's better to throw an error if the records contain different lists of changed fields. Data integrity should always be prioritized; implicitly overriding DBMS's DEFAULT is probably not a good idea. And I think the user would totally understand why there's an error here since the underlying SQL insert statement itself requires a fixed column list.

**Comment by lukaseder** [comment]
> While in terms of API interface, the current design is good, I would argue that it's better to throw an error if the records contain different lists of changed fields.

That sounds like we're about to break thousands of jOOQ applications out there. While I can see how this may have prevented errors in your case, it may well be what people *want* to do. There are many ways to create the same kind of statement. Another way would be this:

```java
dsl.insertInto(TABLE)
   .set(TABLE.FIELD1, "value1")
   .set(TABLE.FIELD2, "value2")
   .newRecord()
   .set(TABLE.FIELD1, "value1")
   .execute();
```

It's quite reasonable to assume a user *explicitly* wanted the current behaviour (especially in the absence of `DEFAULT` expressions), and not have this code throw an exception. I think we should continue to default to this behaviour.

We could add a setting that throws an exception in such a case, but is this really necessary? In the presence of a `NOT NULL` constraint, you'd get the exception anyway.

> Data integrity should always be prioritized

Sure, but you're not getting any integrity problems because of this. Either, you allow nulls and then, the behaviour is again a bit weird but not entirely wrong, or you don't allow nulls using a `NOT NULL` constraint, and the statement fails on the RDBMS level, which is more sound anyway.

> And I think the user would totally understand why there's an error here since the underlying SQL insert statement itself requires a fixed column list.

No user would totally understand why their application suddenly breaks after upgrading jOOQ - and possibly even subtly, because whether they set a value to a record may depend on many things. We cannot change this easily. 

But you can! You can add a trigger that rewrites all inserted `NULL` values to the `DEFAULT` value. Also, you can raise the error yourself by adding a `NOT NULL` constraint.

Anyway. There might be a way to fix this. We could use the generated code's `DEFAULT` expression, which we know and reproduce that in case the `DEFAULT` keyword is not supported. I'll think about it.

Summary:

- In the absence of a `DEFAULT` clause, the current behaviour is correct
- In the presence of a `DEFAULT` clause, the current behaviour is inconsistent with single record and batch inserts, and thus surprising
- A `NOT NULL` constraint would produce the exception you want
- A trigger can work around the status quo
- Perhaps, we can generate `DEFAULT` expressions using the `DEFAULT` keyword, or the available expression from the code generator
- If we don't have this meta data, then we'll stick with the status quo

**Comment by ethanttbui** [comment]
> No user would totally understand why their application suddenly breaks after upgrading jOOQ - and possibly even subtly, because whether they set a value to a record may depend on many things. We cannot change this easily.

Yes, I agree that this behavior cannot be changed easily without backward-compatibility issue. I was not really suggesting that neither (sorry if I made it sound like that), just wondering if we can have an option to customize the behavior.

> Sure, but you're not getting any integrity problems because of this.

You are right. I could've called it an issue with unexpected data or something.

> Perhaps, we can generate `DEFAULT` expressions using the `DEFAULT` keyword, or the available expression from the code generator

If this is doable, it's great. In my application, sometimes a nullable column has a non-null default value, which carries some meaning to the business logic. Writing a trigger probably won't work because we must be able to set it to `null` also. For now, I will manually implement the check and it should be good enough.
Thank you for considering it.

**Comment by lukaseder** [comment]
> just wondering if we can have an option to customize the behavior.

I'd rather not, in this case. There are so many `Settings` already, each of which possibly interacting with other settings and dialects, etc. The more of these we add, the harder it is to integration test things.

Especially, since this particular kind of exception can be raised more reliably using a SQL `NOT NULL` constraint.

> If this is doable, it's great. In my application, sometimes a nullable column has a non-null default value, which carries some meaning to the business logic. I will consider your suggestion of writing a trigger more carefully. Still, it will add some implicit behavior to the app, so I'm not very fond of it.

You're suggesting to add some implicit behaviour to jOOQ - I don't see the difference, at least not to the suggested fix. The trigger is just a workaround until this issue here is fixed. You can name it `trg_jooq_11135` and drop it once this is implemented, if that helps :)

**Comment by ethanttbui** [comment]
The implicit behavior here is that if there's a need for explicitly inserting `null` value and the person is not aware of this implicit behavior, he/she will still be surprised.
But then, in this case, a simple check will be enough for me since I'm writing a base query model that should be used universally in the application.
If a need for a more robust solution arises, I will probably resort to trigger :)
